Implement where conditions for active sessions

codingllama opened the pull request:
Implements [RFD 45 / "where" conditions for active sessions](https://github.com/gravitational/teleport/blob/master/rfd/0045-ssh_session-where-condition.md).

In few words, the purpose of the RFD is to allow the creation of roles that permits users to only join a subset of active sessions (for example, only their own sessions).

Implementation goes a bit further than the RFD, allowing the conditions to be applied to  `update` and `delete` verbs as well.

Originally implemented by @andrejtokarcik (https://github.com/gravitational/teleport/pull/8568), tweaks by @codingllama.

andrejtokarcik committed:
Implement where conditions for active sessions list/read

andrejtokarcik committed:
actionWithConditionForList => actionForListWithCondition

codingllama committed:
Add tests for "where" conditions on active sessions

codingllama committed:
Make Context-exposed sessions follow the RFD API

codingllama committed:
Fix typos

codingllama commented:
Note to reviewers: I've kept Andrej's commits intact, doing the minimum required tweaks to get up-to-spec and tested in separate commits.

I have plans for a few follow ups, but it seemed better to get the bulk of the feature in and refine later. In its final stage we intend to backport v8.

Issue https://github.com/gravitational/teleport/issues/5430.

quinqu commented on lib/services/parser.go:
```diff
@@ -232,11 +237,59 @@ func (ctx *Context) GetIdentifier(fields []string) (interface{}, error) {
+	ServerAddr string `json:"server_addr"`
+	// ClusterName is the name of cluster that this session belongs to.
+	ClusterName string `json:"cluster_name"`
+
```
Extra line

quinqu commented on lib/auth/auth_with_roles.go:
```diff
@@ -2039,21 +2067,40 @@ func (a *ServerWithRoles) UploadSessionRecording(r events.SessionRecording) erro
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, ef := range sessionEvents {
```
I think here you can use `SearchEvents()` and pass in `session.end` string value in eventTypes ([]string) instead of iterating through all events. There should only be one `session.end` event per session.

quinqu commented on lib/events/filelog.go:
```diff
@@ -374,7 +374,7 @@ func (l *FileLog) SearchSessionEvents(fromUTC, toUTC time.Time, limit int, order
+	condition  utils.FieldsCondition
 }
 
 // Close closes the audit log, which inluces closing all file handles and releasing
```
I know you didn't make this typo

```suggestion
// Close closes the audit log, which includes closing all file handles and releasing
```

codingllama replied:
Done, not to worry.

So many typos on the codebase, even with misspell. I wonder if there's a better automated checker somewhere else.

quinqu commented on lib/services/parser.go:
```diff
@@ -232,11 +237,59 @@ func (ctx *Context) GetIdentifier(fields []string) (interface{}, error) {
+
+	// Parties is a list of session participants expressed as usernames.
+	// Changed from a []session.Party in the original session.Session.
+	Parties []string `json:"parties"`
```
I think it would be more intuitive to call this `SessionUsernames` or `Usernames`. If I saw this w/o context, I would think `Parties` is of type `[]*Party`.

codingllama replied:
I'm following the RFD here, but personally I do agree with you. `Usernames` seems a good call; [RFD 44](https://github.com/gravitational/teleport/blob/master/rfd/0044-session-where-condition.md) uses `session.participants`, which seems to fit well here too.

WDYT? I'm tempted to go with `Participants` so both session objects match.

rosstimothy replied:
+1 for `Participants`

codingllama replied:
Done.

quinqu commented on lib/session/session.go:
```diff
@@ -108,6 +109,15 @@ type Session struct {
 	ClusterName string `json:"cluster_name"`
 }
 
+// Users returns a list of users involved in this session as parties.
```
```suggestion
// Users returns a list of users involved in this session as strings.
```

codingllama replied:
Done, well spotted.

alex-kovoy commented on lib/session/session.go:
```diff
@@ -276,20 +286,43 @@ func activeKey(namespace string, key string) []byte {
 // if no sessions are active
-func (s *server) GetSessions(namespace string) ([]Session, error) {
-	prefix := activePrefix(namespace)
+func (s *server) GetSessions(namespace string, cond *types.WhereExpr) ([]Session, error) {
```
I would refrain from introducing nullable parameters in the API method signature as it makes a method call no longer self-descriptive. Since we are changing input parameters anyway, it might be a good time to introduce  `GetSessionRequest`  (also more scallable)

codingllama replied:
Agreed - I didn't want to change too much of the original commits, but this struck out to me as well. Having `cond` as a param seems to make it easy to (inadvertently) skip permission checks too.

Jane had similar comments as well - I'll do a few changes and let you know when ready.

codingllama replied:
Done.

alex-kovoy commented on lib/events/firestoreevents/firestoreevents.go:
```diff
@@ -630,7 +630,7 @@ func (l *Log) searchEventsOnce(fromUTC, toUTC time.Time, namespace string, limit
 	filter := searchEventsFilter{eventTypes: []string{events.SessionEndEvent}}
 	if cond != nil {
-		condFn, err := events.ToEventFieldsCondition(cond)
+		condFn, err := utils.ToFieldsCondition(cond)
```
```golang
SearchSessionEvents(fromUTC, toUTC time.Time, limit int, order types.EventOrder, startKey string, cond *types.WhereExpr)
```

Another candidate to refactor method parameters to a request object

codingllama replied:
Yep, it's a mouthful.

I'll take a look, but I won't make promises for this particular one - idk how much trouble it might be to change it and it's not strictly part of the PR.

codingllama replied:
Added a TODO to follow up in a separate PR - I'd like to keep the scope here somewhat contained. If folks like the final result of GetSessions then we can go for something similar here.

codingllama commented on lib/web/apiserver.go:
```diff
@@ -1970,7 +1970,7 @@ func (h *Handler) siteSessionsGet(w http.ResponseWriter, r *http.Request, p http
 	}
 
-	sessions, err := clt.GetSessions(namespace)
+	sessions, err := clt.GetSessions(namespace, nil)
```
Yes, that struck out to me too. I'll give it a try.

codingllama replied:
Done.

codingllama commented:
Thanks Jane and Alexey for the prompt review!

codingllama committed:
Fix typos and spacing

rosstimothy commented on lib/utils/fields.go:
```diff
@@ -0,0 +1,156 @@
+	if !found {
+		return ""
+	}
+	v, _ := val.(string)
```
Why are we ignoring `ok` here but not in the other `GetX` methods? If we really don't need to check it here maybe just use `v := val.(string)`

codingllama replied:
Changed to match your suggestion.

codingllama committed:
Rename "parties" to "participants" in the context session

codingllama committed:
Update RFD to reflect PR changes

Update RFD to reflect PR changes

Specifically, mark as implemented and rename `parties` to
`participants`.

codingllama committed:
Push list authz logic to ServerWithRoles, obsolete cond

codingllama committed:
Remove cond from GetSessions signature

codingllama committed:
Simplify cast in lib.utils.Fields.GetString

codingllama committed:
Add TODO to refactor SearchSessionEvents / stored sessions

codingllama commented:
PTAL? Rebased and addressed all comments.
